### ISSUE TITLE
Controlling listener at runtime

### ISSUE THREAD

**Issue by venkateshatwork** [spring-projects/spring-kafka#370]
Team,

This is a clarification, not an issue. I would like to control Kafka listener at runtime. Here is the below usecase. Am sure there must be a way to do this, but unable to find it.

1. 10 messages are sent to a topic A from producer
2. In consumer side, I do not want to consume all 10 messages instantly. I want to process first message, after the process is complete, go for the next message.
3. Is there a way to tell listener to wait for the first message to complete processing and go for next ? Messages should not be lost.

**Comment by artembilan** [comment]
Your expectations are unclear. 
That is how it works by default.

The next message doesn't go to the listener until that hasn't finished withe the previous:
```
private void invokeRecordListener(final ConsumerRecords<K, V> records) {
			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
			while (iterator.hasNext()) {
				final ConsumerRecord<K, V> record = iterator.next();
				if (this.logger.isTraceEnabled()) {
					this.logger.trace("Processing " + record);
				}
				try {
					switch (this.listenerType) {
						case ACKNOWLEDGING_CONSUMER_AWARE:
							this.listener.onMessage(record,
									this.isAnyManualAck
											? new ConsumerAcknowledgment(record)
											: null, this.consumer);
							break;
```
You see, we iterate over a batch returned from the `Consumer.poll()` and call `listener` until we have records. 

The messages in Kafka can't be lost by the definition. That is just consumer group offset management which makes us to move in the sequence of records in Kafka logs. 

The offset commit is controlled by the particular container options: http://docs.spring.io/spring-kafka/reference/html/_reference.html#committing-offsets

But in any case the messages isn't lost. And if one of the records in the batch causes error all the rest won't be processed and the next `poll()`return the batch starting from the previous successfully commited offset.

**Comment by venkateshatwork** [comment]
Let me rephrase in different way. The method invokeRecordListener(..) pulls all records from the broker. I want to configure such a way that pull 1 record, let listener process the record and go for next. The current implementation pulls all the records and process one by one, but I want to restrict such a way that consumer should pull one message at a time, finish processing it and pull the next from the broker.

**Comment by garyrussell** [comment]
It's not a very efficient way to use kafka but you can set `max.poll.records` to `1` - [see the kafka documentation](https://kafka.apache.org/documentation/).

The constant in `ConsumerConfig` is

      public static final String MAX_POLL_RECORDS_CONFIG = "max.poll.records";
      private static final String MAX_POLL_RECORDS_DOC = "The maximum number of records returned in a single call to poll().";

With spring-kafka 1.2 and earlier; we will fetch another record in the background; with the upcoming 1.3 release the listener is called on the consumer thread so there will always only be the fetched record locally.

**Comment by venkateshatwork** [comment]
This is cool stuff, let me try it, I agree it is not efficient way to use Kafka, but Kafka is used extensively in our application for the purpose it meant for. But one of the feature need to control the messages polled, instead of going to different implementation, I can have dedicated topic and consumer with poll 1 to achieve this.

**Comment by artembilan** [comment]
I think we can close this issue with the feature provided in the Spring Integration Kafka via `KafkaMessageSource`: https://github.com/spring-projects/spring-integration-kafka/issues/184.
